### PR TITLE
issue 1111 only initialize Realm and KeyChainService once

### LINKED ISSUE #1111: realm crash: Realm file decryption failed
I'm still getting this and have to reset storage to continue using the app.

The interesting thing is that it seemed to happen for both apps at the same time this morning. So I wonder, maybe the encryption key we use from keychain is not stable? 

```
Unable to open a realm at path '/var/mobile/Containers/Data/Application/850D5417-D347-4FCC-BDB4-A1AE3E1A2200/Documents/encrypted.realm': Realm file decryption failed Path:Exception backtrace:
0   FlowCrypt                           0x00000001049df9fc _ZN5realm4util6detail26ExceptionWithBacktraceBaseC2Ev + 64
1   FlowCrypt                           0x00000001049ee55c _ZN5realm4util22ExceptionWithBacktraceISt13runtime_errorEC2IJRKNSt3__112basic_stringIcNS5_11char_traitsIcEENS5_9allocatorIcEEEEEEEDpOT_ + 104
2   FlowCrypt                           0x00000001049ee310 _ZN5realm4util4File11AccessErrorC2ERKNSt3__112basic_stringIcNS3_11char_traitsIcEENS3_9allocatorIcEEEESB_ + 84
3   FlowCrypt                           0x00000001049ee29c _ZN5realm15InvalidDatabaseC2ERKNSt3__112basic_stringIcNS1_11char_traitsIcEENS1_9allocatorIcEEEES9_ + 88
4   FlowCrypt                           0x00000001049e80ec _ZN5realm15InvalidDatabaseC1ERKNSt3__112basic_stringIcNS1_11char_traitsIcEENS1_9allocatorIcEEEES9_ + 44
5   FlowCrypt                           0x00000001049e79fc _ZN5realm9SlabAlloc11attach_fileERKNSt3__112basic_stringIcNS1_11char_traitsIcEENS1_9allocatorIcEEEERNS0_6ConfigE + 4156
6   FlowCrypt                           0x0000000104aabf14 _ZN5realm2DB4openERKNSt3__112basic_stringIcNS1_11char_traitsIcEENS1_9allocatorIcEEEEbNS_9DBOptionsE + 4028
7   FlowCrypt                           0x0000000104aaf544 _ZN5realm2DB4openERNS_11ReplicationERKNSt3__112basic_stringIcNS3_11char_traitsIcEENS3_9allocatorIcEEEENS_9DBOptionsE + 244
8   FlowCrypt                           0x0000000104ab7124 _ZN5realm2DB6createENSt3__110unique_ptrINS_11ReplicationENS1_14default_deleteIS3_EEEERKNS1_12basic_stringIcNS1_11char_traitsIcEENS1_9allocatorIcEEEENS_9DBOptionsE + 324
9   FlowCrypt                           0x0000000104cc79f0 _ZN5realm5_impl16RealmCoordinator7open_dbEv + 1444
10  FlowCrypt                           0x0000000104cc963c _ZN5realm5_impl16RealmCoordinator12do_get_realmENS_5Realm6ConfigERNSt3__110shared_ptrIS2_EENS_4util8OptionalINS_9VersionIDEEERNS8_17CheckedUniqueLockE + 92
11  FlowCrypt                           0x0000000104cc9444 _ZN5realm5_impl16RealmCoordinator9get_realmENS_5Realm6ConfigENS_4util8OptionalINS_9VersionIDEEE + 692
12  FlowCrypt                           0x0000000104e73204 _ZN5realm5Realm16get_shared_realmENS0_6ConfigE + 156
13  FlowCrypt                           0x00000001049498f8 +[RLMRealm realmWithConfiguration:queue:error:] + 1968
14  FlowCrypt                           0x00000001055aa4e4 $sSo8RLMRealmC13configuration5queueABSo0A13ConfigurationC_So012OS_dispatch_C0CSgtKcfCTO + 128
15  FlowCrypt                           0x000000010564cba8 $s10RealmSwift0A0V13configuration5queueA2C13ConfigurationV_So012OS_dispatch_D0CSgtKcfC + 200
16  FlowCrypt                           0x00000001045b4328 $s9FlowCrypt16EncryptedStorageC8validateyyKF + 276
17  FlowCrypt                           0x000000010447d544 $s9FlowCrypt12GlobalRouterC24validateEncryptedStorage33_57D5652C71D18AB9D7F3E85674F97089LLyyyyXEF + 156
18  FlowCrypt                           0x000000010447d010 $s9FlowCrypt12GlobalRouterC7proceedyyF + 64
19  FlowCrypt                           0x0000000104372150 $s9FlowCrypt11AppDelegateC11application_29didFinishLaunchingWithOptionsSbSo13UIApplicationC_SDySo0k6LaunchJ3KeyaypGSgtF + 316
20  FlowCrypt                           0x0000000104372240 $s9FlowCrypt11AppDelegateC11application_29didFinishLaunchingWithOptionsSbSo13UIApplicationC_SDySo0k6LaunchJ3KeyaypGSgtFTo + 184
21  UIKitCore                           0x000000018326b044 8388EB03-002B-3B35-A78A-6A022894292E + 3313732
22  UIKitCore                           0x0000000183452b38 8388EB03-002B-3B35-A78A-6A022894292E + 5311288
23  UIKitCore                           0x000000018343b4a4 8388EB03-002B-3B35-A78A-6A022894292E + 5215396
24  UIKitCore                           0x0000000183298e44 8388EB03-002B-3B35-A78A-6A022894292E + 3501636
25  UIKitCore                           0x00000001830f162c _UIScenePerformActionsWithLifecycleActionMask + 104
26  UIKitCore                           0x000000018321cb7c 8388EB03-002B-3B35-A78A-6A022894292E + 2993020
27  UIKitCore                           0x000000018333d380 8388EB03-002B-3B35-A78A-6A022894292E + 4174720
28  UIKitCore                           0x000000018357df8c 8388EB03-002B-3B35-A78A-6A022894292E + 6537100
29  UIKitCore                           0x0000000183509710 8388EB03-002B-3B35-A78A-6A022894292E + 6059792
30  UIKitCore                           0x00000001830f2430 8388EB03-002B-3B35-A78A-6A022894292E + 1770544
31  UIKitCore                           0x00000001831b2350 8388EB03-002B-3B35-A78A-6A022894292E + 2556752
32  UIKitCore                           0x00000001830f4340 _UISceneSettingsDiffActionPerformChangesWithTransitionContext + 276
33  UIKitCore                           0x00000001831dedf4 8388EB03-002B-3B35-A78A-6A022894292E + 2739700
34  UIKitCore                           0x0000000183666260 8388EB03-002B-3B35-A78A-6A022894292E + 7488096
35  UIKitCore                           0x0000000183122c60 8388EB03-002B-3B35-A78A-6A022894292E + 1969248
36  UIKitCore                           0x0000000183166524 8388EB03-002B-3B35-A78A-6A022894292E + 2245924
37  UIKitCore                           0x000000018329dd70 8388EB03-002B-3B35-A78A-6A022894292E + 3521904
38  UIKitCore                           0x00000001831dd0b4 8388EB03-002B-3B35-A78A-6A022894292E + 2732212
39  FrontBoardServices                  0x000000019262ce20 6FDEC0C3-E500-3E06-948E-E34BDA520801 + 171552
40  FrontBoardServices                  0x0000000192652cdc 6FDEC0C3-E500-3E06-948E-E34BDA520801 + 326876
41  FrontBoardServices                  0x000000019260d6b4 6FDEC0C3-E500-3E06-948E-E34BDA520801 + 42676
42  FrontBoardServices                  0x000000019260ecf4 6FDEC0C3-E500-3E06-948E-E34BDA520801 + 48372
43  libdispatch.dylib                   0x00000001807a7660 A5CBAAB3-E389-3548-BAAC-FAB18411B94A + 18016
44  libdispatch.dylib                   0x00000001807ab118 A5CBAAB3-E389-3548-BAAC-FAB18411B94A + 33048
45  FrontBoardServices                  0x000000019260ef94 6FDEC0C3-E500-3E06-948E-E34BDA520801 + 49044
46  FrontBoardServices                  0x000000019260e3d4 6FDEC0C3-E500-3E06-948E-E34BDA520801 + 46036
47  FrontBoardServices                  0x00000001926129e4 6FDEC0C3-E500-3E06-948E-E34BDA520801 + 63972
48  CoreFoundation                      0x0000000180b57020 B2D21CFD-378C-36D5-BAF7-3F70599CFEFC + 765984
49  CoreFoundation                      0x0000000180b67ce0 B2D21CFD-378C-36D5-BAF7-3F70599CFEFC + 834784
50  CoreFoundation                      0x0000000180aa2054 B2D21CFD-378C-36D5-BAF7-3F70599CFEFC + 24660
51  CoreFoundation                      0x0000000180aa77f4 B2D21CFD-378C-36D5-BAF7-3F70599CFEFC + 47092
52  CoreFoundation                      0x0000000180abb3b8 CFRunLoopRunSpecific + 600
53  GraphicsServices                    0x000000019c44b38c GSEventRunModal + 164
54  UIKitCore                           0x000000018345b6a8 8388EB03-002B-3B35-A78A-6A022894292E + 5346984
55  UIKitCore                           0x00000001831da7f4 UIApplicationMain + 2092
56  FlowCrypt                           0x00000001043ba1bc $s9FlowCryptyyXEfU_ + 364
57  FlowCrypt                           0x00000001043ba218 $ss5Error_pIgzo_ytsAA_pIegrzo_TR + 24
58  FlowCrypt                           0x00000001043ba26c $ss5Error_pIgzo_ytsAA_pIegrzo_TRTA + 28
59  libswiftObjectiveC.dylib            0x00000001ae902a20 $s10ObjectiveC15autoreleasepool8invokingxxyKXE_tKlF + 64
60  FlowCrypt                           0x00000001043ba034 main + 88
61  dyld                                0x000000010b539a24 start + 520.


```

### PR DESCRIPTION
**Pull request by tomholub**
This PR cleans up realm init to minimize race conditions and have more control over when and how it's initialized.

close #1111 

WIP

----------------------------------

- Does not need tests (refactor only, docs or internal changes)

--------------------------------

### To be filled by reviewers

I have reviewed that this PR... _(tick whichever items you personally focused on during this review)_:
- [ ] addresses the issue it closes (if any)
- [ ] code is readable and understandable
- [ ] is accompanied with tests, or tests are not needed
- [ ] is free of vulnerabilities
- [ ] is documented clearly and usefully, or doesn't need documentation

### CODE REVIEW

**Comment by tomholub** [comment]
Sorry for another monstrous PR. The main goal of this PR was:
 - initialize Realm only once, and in a controlled manner
 - make the initialization a MainActor
 - pass initialized storage to places that need it (which turned out to be every single place of the app 😞  ) through dependency injection

For this I created `AppContext` that gets passed around.

While I was doing it, certain things kept getting in my way, in particular magical `currentUser` usage hidden inside classes/methods, which needed `DataProvider` which needed `EncryptedStorage` which needed that initialized Realm.. it was a mess, and so I also did some more work:
  - where the above situation was particularly upsetting, I've instead added `currentUser` as method or class parameter
  - I've done some prepwork for #863 by requiring user to be present early in VC init. Next step, for that issue, would be to pass `User` into each VC init that requires it.

While doing this, I noticed some classes are used across the app and rely on storage, and so I've added them into `AppContext` as well. Which has gotten bigger than I expected.

Things that bother me about this PR and will need to get cleaned up (hopefully in other PRs):
 - the dependency injection highlights an overall mess present across the app. There are too many services, providers, often named wrongly, often their constructor properties are named wrongly, etc. It was more hidden until now when you could just put `dataProvider = DataProvider()` in every constructor, but now that it's truly dependency-injected, it really badly brings out some ugliness that we need to work on
 - repetitive code. I have written some code sequences 5+times again and again while refactoring this, and it would need a second look
 - general verbosity. similar as point above, and needs to be improved
 - inconsitency. Sometimes it is `user`, sometimes `activeUser`, sometimes `currentUser`.. some abstractions are somewhat different on IMAP and Gmail, and need to be alighned. etc

As of now, the app builds and runs and I didn't run into a bug yet. But Swift tests are likely broken. I'll have a look.

File: FlowCrypt/App/AppContext.swift [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+                dataService: dataService
+            ),
+            dataService: dataService,
+            keyStorage: keyStorage,
+            keyService: keyService,
+            passPhraseService: passPhraseService,
+            clientConfigurationService: clientConfigurationService,
+            globalRouter: globalRouter
+        )
+    }
```
Comment (tomholub) [inline FlowCrypt/App/AppContext.swift:80]: This is the main part of the PR

File: FlowCrypt/Controllers/CheckMailAuth/CheckMailAuthViewController.swift [hunk 2]
```diff
@@ -13,8 +13,10 @@ import UIKit
...
 class CheckMailAuthViewController: TableNodeViewController {
     private let globalRouter: GlobalRouterType
+    private let appContext: AppContext
 
-    init(globalRouter: GlobalRouterType = GlobalRouter()) {
+    init(appContext: AppContext, globalRouter: GlobalRouterType = GlobalRouter()) {
+        self.appContext = appContext
         self.globalRouter = globalRouter
         super.init(node: TableNode())
     }
```
Comment (tomholub) [inline FlowCrypt/Controllers/CheckMailAuth/CheckMailAuthViewController.swift:22]: Context now gets passed as first argument of every VC init.

Next step will be to pass `User` (not optional) into every VC that requires it. #863

File: FlowCrypt/Controllers/Setup/SetupInitialViewController.swift [hunk 3]
```diff
@@ -57,25 +59,31 @@ final class SetupInitialViewController: TableNodeViewController {
...
-        clientConfigurationService: ClientConfigurationServiceType = ClientConfigurationService(),
-        emailKeyManagerApi: EmailKeyManagerApiType = EmailKeyManagerApi()
+        emailKeyManagerApi: EmailKeyManagerApiType? = nil
     ) {
+        self.appContext = appContext
         self.user = user
+        let backupService = backupService ?? BackupService(
+            backupProvider: appContext.getRequiredMailProvider().backupProvider,
+            messageSender: appContext.getRequiredMailProvider().messageSender
+        )
```
Comment (tomholub) [inline FlowCrypt/Controllers/Setup/SetupInitialViewController.swift:79]: Notice how backupService is initialized with nil first and only then initialized properly. This is an "optional parameter" and it's done this way because the initialization needs another parameter which cannot be used directly above.

I've written this pattern probably 100x in this PR and it bothers me.

It's only ever needed if we wanted to unit-test the VC but we don't really unit test VCs. So with a bit of luck, I may rip all of this out and make most VC parameters required and supplied from context or from whoever calls the VC. That should clean things up.

File: FlowCrypt/Controllers/SideMenu/Menu/MyMenuViewController.swift [hunk 4]
```diff
@@ -64,14 +65,21 @@ final class MyMenuViewController: ASDKViewController<ASDisplayNode> {
...
-        self.dataService = dataService
+        guard let currentUser = appContext.dataService.currentUser else {
+            fatalError("no current user") // todo - dependency-inject
+        }
+        self.currentUser = currentUser
+        self.appContext = appContext
+        self.foldersProvider = foldersProvider ?? FoldersService(
+            encryptedStorage: appContext.encryptedStorage,
+            remoteFoldersProvider: appContext.getRequiredMailProvider().remoteFoldersProvider
+        )
```
Comment (tomholub) [inline FlowCrypt/Controllers/SideMenu/Menu/MyMenuViewController.swift:82]: I initialized `FolderService` many times this way, good candidate to go into `AppContext` (a method, maybe, similar to `EmailProvider`)

File: FlowCrypt/Functionality/DataManager/DataService.swift [hunk 5]
```diff
@@ -125,34 +120,3 @@ extension DataService: DBMigration {
         try await migrationService.performMigrationIfNeeded()
     }
 }
-
-// MARK: - SessionProvider
-extension DataService: ImapSessionProvider {
-    func imapSession() -> IMAPSession? {
```
Comment (tomholub) [inline FlowCrypt/Functionality/DataManager/DataService.swift:131]: Moved to its own file, no need to be here

File: FlowCrypt/Functionality/DataManager/Encrypted Storage/KeyChainService.swift [hunk 6]
```diff
@@ -10,64 +10,60 @@ import FlowCryptCommon
...
+actor KeyChainService {
 
-protocol KeyChainServiceType {
-    func getStorageEncryptionKey() throws -> Data
-}
-
-struct KeyChainService: KeyChainServiceType {
-    private static var logger = Logger.nested(in: Self.self, with: "Keychain")
+    private let logger = Logger.nested("KeyChain")
+    private let keyByteLen = 64
```
Comment (tomholub) [inline FlowCrypt/Functionality/DataManager/Encrypted Storage/KeyChainService.swift:20]: This class was rewritten without changing functionality. Just general cleanup (used to be somewhat static). 

Now an actor.

File: FlowCrypt/Functionality/Mail Provider/Gmail/GmailService.swift [hunk 7]
```diff
@@ -35,10 +35,11 @@ class GmailService: MailServiceProvider {
...
     init(
-        userService: GoogleUserServiceType = GoogleUserService(),
+        currentUserEmail: String,
+        userService: GoogleUserServiceType? = nil,
         backupSearchQueryProvider: GmailBackupSearchQueryProviderType = GmailBackupSearchQueryProvider()
     ) {
-        self.userService = userService
+        self.userService = userService ?? GoogleUserService(currentUserEmail: currentUserEmail)
         self.backupSearchQueryProvider = backupSearchQueryProvider
     }
```
Comment (tomholub) [inline FlowCrypt/Functionality/Mail Provider/Gmail/GmailService.swift:44]: Again example of the `nil ?? Something()` pattern I don't like much.

File: FlowCrypt/Functionality/Mail Provider/MailProvider.swift [hunk 8]
```diff
@@ -81,12 +76,12 @@ final class MailProvider {
...
-        currentAuthType: @autoclosure @escaping () -> (AuthType?),
-        services: [MailServiceProvider]
+    init(
+        currentAuthType: AuthType,
+        currentUser: User
     ) {
         self.currentAuthType = currentAuthType
-        self.services = services
+        self.services = MailServiceProviderFactory.services(user: currentUser)
     }
```
Comment (tomholub) [inline FlowCrypt/Functionality/Mail Provider/MailProvider.swift:85]: I was a bit worried about this - originally dynamic as a closure, now user is directly provided. Need to scrutinize more. But app seems to work the same.

File: FlowCrypt/Functionality/Services/GlobalRouter.swift [hunk 9]
```diff
@@ -36,75 +36,76 @@ final class GlobalRouter {
...
+            do {
+                try appContext.encryptedStorage.validate()
+                proceed(with: appContext)
+            } catch {
+                renderInvalidStorageView(error: error, encryptedStorage: nil)
+            }
+        } catch {
+            renderInvalidStorageView(error: error, encryptedStorage: nil)
         }
     }
```
Comment (tomholub) [inline FlowCrypt/Functionality/Services/GlobalRouter.swift:59]: Main part of the PR. Set up context first, then pass it around. Don't re-setup everything in many places.

**Review by tomholub — COMMENTED** [review]
Some highlights below

**Comment by ivan-ushakov** [comment]
Am I right that we have now one global context with app lifetime? What about Realm and threading? I see that context is forced to be created on UI thread, so Realms is created on that thread. That means we must do any read query on UI thread.

**Comment by tomholub** [comment]
> Am I right that we have now one global context with app lifetime? What about Realm and threading? I see that context is forced to be created on UI thread, so Realms is created on that thread. That means we must do any read query on UI thread.

That can be reworked - I don't mind to create Realm on any thread. 

But the encryption key for realm, which uses KeyChain, must be created on main thread, to be safe.

I don't exactly mind sticking to main thread for Realm either, though.

**Comment by tomholub** [comment]
@ivan-ushakov could you please help me with these when running tests. I'm not very familiar with this stack.

image

**Comment by ivan-ushakov** [comment]
> @ivan-ushakov could you please help me with these when running tests. I'm not very familiar with this stack.

Sure, let me check.

**Comment by tomholub** [comment]
I'm still hoarding some changes here. I'll go ahead and push them

File: FlowCrypt/App/main.swift [hunk 10]
```diff
@@ -9,16 +9,9 @@
 import Foundation
 import UIKit
 
-autoreleasepool {
```
Comment (ivan-ushakov) [inline FlowCrypt/App/main.swift:12]: Not sure we could skip `autoreleasepool` here

**Comment by ivan-ushakov** [comment]
> I'm still hoarding some changes here. I'll go ahead and push them

Looks like you added some UI classes to unit test target `FlowCryptAppTests`. We don't compile UI classes for unit tests so we don't add some Pod dependencies. I guess that this is the reason.

File: FlowCrypt/App/main.swift (hunk 10 above)
Comment (tomholub) [reply FlowCrypt/App/main.swift:12]: I'm not sure either - the IDE was telling me that it's no longer needed. @sosnovsky may know

**Comment by tomholub** [comment]
> > I'm still hoarding some changes here. I'll go ahead and push them
> 
> Looks like you added some UI classes to unit test target `FlowCryptAppTests`. We don't compile UI classes for unit tests so we don't add some Pod dependencies. I guess that this is the reason.
> 
> You could check `Target Membership` on the right side panel and remove checkmark for `FlowCryptAppTests` on UI files.

Actually, the opposite happened at first - every single app file was removed from test targets (even the ones that were there previously, like `Core` and many others). So I've added them back. I'm aware that I shouldn't be adding controllers and other UI classes to tests, but after my refactor it seemed hard to find the right boundary of what to include and what not in a way that the build process wouldn't break. That should be improved, for sure.

But I don't think that's the cause of it. For example, a bunch of the messages are from `SwiftyRSA` which is needed by `Core` which was always tested before, and always in test target.

Could you look at `SwiftyRSA` specifically first?

image

**Comment by tomholub** [comment]
Some good news - at least appium tests are passing

image

**Comment by ivan-ushakov** [comment]
Should work now

File: FlowCrypt/Functionality/Services/AppStartup.swift [hunk 11]
```diff
@@ -11,27 +11,33 @@ import UIKit
 
 private let logger = Logger.nested("AppStart")
 
-struct AppStartup {
+class AppStartup {
```
Comment (sosnovsky) [inline FlowCrypt/Functionality/Services/AppStartup.swift:14]: I think we can leave it as `struct`, no need to use `class` here

**Comment by tomholub** [comment]
image

I'd still like to be using custom arguments in the future - for appium tests mocks

File: FlowCrypt/Controllers/MessageList Extension/MsgListViewController.swift [hunk 12]
```diff
@@ -10,47 +10,43 @@ import UIKit
 
 @MainActor
 protocol MsgListViewController {
-    func open(with message: InboxRenderable, path: String)
+    func open(appContext: AppContext, with message: InboxRenderable, path: String)
```
[the remark this answers is not in the document]
Comment (tomholub) [reply FlowCrypt/Controllers/MessageList Extension/MsgListViewController.swift:13]: agree

File: FlowCrypt/Controllers/Settings/Backup/Backups Option Scene/BackupOptionsViewController.swift [hunk 13]
```diff
@@ -35,17 +35,24 @@ final class BackupOptionsViewController: ASDKViewController<TableNode> {
...
         userId: UserId
     ) {
+        self.appContext = appContext
         self.decorator = decorator
         self.backups = backups
-        self.backupService = backupService
+        let backupService = backupService ?? BackupService(
+            backupProvider: appContext.getRequiredMailProvider().backupProvider,
+            messageSender: appContext.getRequiredMailProvider().messageSender
+        )
```
[the remark this answers is not in the document]
Comment (tomholub) [reply FlowCrypt/Controllers/Settings/Backup/Backups Option Scene/BackupOptionsViewController.swift:53]: That would be great. But I have found that this may mean we would need to include `AppContext` in test target which is not desirable (because then everything gets included from there).

So I'm thinking I may instead add `getBackupProvider()` method to AppContext. It's less clean though.

I'll fiddle with it some more, see if I can add AppContext without adding everything (router / delegate / and with that all VCs.. not desired in tests)

File: FlowCrypt/Controllers/Setup/SetupGenerateKeyViewController.swift [hunk 14]
```diff
@@ -182,10 +178,10 @@ private actor Service {
     }
 
     private func getUserId() throws -> UserId {
-        guard let email = storage.email, !email.isEmpty else {
+        guard let email = appContext.dataService.email, !email.isEmpty else {
             throw CreateKeyError.missedUserEmail
         }
-        guard let name = storage.email, !name.isEmpty else {
+        guard let name = appContext.dataService.email, !name.isEmpty else {
```
[the remark this answers is not in the document]
Comment (tomholub) [reply FlowCrypt/Controllers/Setup/SetupGenerateKeyViewController.swift:172]: You really notice things well :-)

**Comment by tomholub** [comment]
@ivan-ushakov I managed to break it again. Although this time there aren't so many files in test target.

What steps did you take to fix this last time? So that I can learn it too.

image

File: FlowCrypt/Controllers/Settings/Backup/Backups Option Scene/BackupOptionsViewController.swift (hunk 13 above)
[the remark this answers is not in the document]
Comment (tomholub) [reply FlowCrypt/Controllers/Settings/Backup/Backups Option Scene/BackupOptionsViewController.swift]: See here https://github.com/FlowCrypt/flowcrypt-ios/pull/1117/commits/e786059bb5492fe8f6d6d686df5871cee185fdc3

I know it's not what you wanted, but I'm unsure how to achieve what you proposed while keeping test target happy. We could file it as another issue, as what you proposed is nicer.

**Comment by ivan-ushakov** [comment]
> @ivan-ushakov I managed to break it again. Although this time there aren't so many files in test target.
> 
> What steps did you take to fix this last time? So that I can learn it too.

1. `git checkout master FlowCrypt.xcodeproj/project.pbxproj`
2. Fix several errors like `missing files`, `links to removed files`. Xcode will help here.

**Comment by sosnovsky** [comment]
> image
> 
> I'd still like to be using custom arguments in the future - for appium tests mocks

These arguments can be parsed in `AppDelegate` using `CommandLine.arguments` array, no need for custom `main.swift`

File: FlowCrypt/Controllers/Settings/Backup/Backups Option Scene/BackupOptionsViewController.swift (hunk 13 above)
[the remark this answers is not in the document]
Comment (sosnovsky) [reply FlowCrypt/Controllers/Settings/Backup/Backups Option Scene/BackupOptionsViewController.swift]: Updated code looks better 👍
Another possible solution is to pass only `MailProvider` (as it's already included in test target), not full `AppContext`:
```swift
init(mailProvider: MailProvider) {
    self.backupProvider = mailProvider.backupProvider
    self.messageSender = mailProvider.messageSender
}
```
But I think your current solution is good enough.

**Comment by tomholub** [comment]
> These arguments can be parsed in AppDelegate using CommandLine.arguments array, no need for custom main.swift

Got it! Filed https://github.com/FlowCrypt/flowcrypt-ios/issues/1130

**Comment by sosnovsky** [comment]
> @ivan-ushakov I managed to break it again. Although this time there aren't so many files in test target.
> 
> What steps did you take to fix this last time? So that I can learn it too.
> 
> image

Current issues are caused by https://github.com/FlowCrypt/flowcrypt-ios/blob/issue-1111-realm-keychain-init/FlowCryptAppTests/Functionality/Services/ComposeMessageServiceTests.swift#L31 - this code needs `CoreHost` which imports `SwiftyRSA` framework, but it's not included in test target.

**Comment by sosnovsky** [comment]
I tried to remove `CoreHost` there, but then I got error about missing `CoreHost` in `Core.swift` file, trying to fix it now

**Comment by tomholub** [comment]
The Core parts are needed for tests in either case. And so is SwiftyRSA.

**Comment by tomholub** [comment]
Thanks for help!

**Comment by sosnovsky** [comment]
Also for testing we can use `in-memory` Realm instance, instead of `on-disk` - it should be faster and doesn't need to be resetted, as it's new instance for each new test (https://docs.mongodb.com/realm/sdk/ios/test-and-debug/#testing)

**Comment by tomholub** [comment]
I'm not entirely sure we use Realm in Swift tests. But will file an issue.

**Comment by sosnovsky** [comment]
> The Core parts are needed for tests in either case. And so is SwiftyRSA.

Yes, this `Core.swift` code is 1 month old and previously worked well, so it's strange why it doesn't run after the latest updates.

**Comment by sosnovsky** [comment]
> I'm not entirely sure we use Realm in Swift tests. But will file an issue.

It's initiated only in `ComposeMessageServiceTests` - https://github.com/FlowCrypt/flowcrypt-ios/blob/issue-1111-realm-keychain-init/FlowCryptAppTests/Functionality/Services/ComposeMessageServiceTests.swift#L36
But maybe later we'll need it in other places too

**Comment by sosnovsky** [comment]
Currently unit tests are running, but `ComposeMessageServiceTests` fail, probably because of some `Realm` issues in test environment (as previously `ComposeMessageService` didn't have `encryptedStorage` in `init`)

**Comment by tomholub** [comment]
> Currently unit tests are running, but `ComposeMessageServiceTests` fail, probably because of some `Realm` issues in test environment (as previously `ComposeMessageService` didn't have `encryptedStorage` in `init`)

Yes but one of its dependencies must have previously still needed Realm, even if Realm wasn't touched in the tests.

Do you want to look into this? I'm ok to temporarily disable those tests, too.

**Comment by sosnovsky** [comment]
All tests are passing now ✅

**Review by sosnovsky — APPROVED** [review]

**Comment by tomholub** [comment]
I'll merge this. After that, I'll review the already merged PR - there are many opportunities for small fixes. Then can settle that after. Thanks!

File: FlowCrypt/Functionality/DataManager/SessionService.swift [hunk 15]
```diff
@@ -9,68 +9,67 @@
...
 
-final class UserAccountService {
+final class SessionService {
     private let encryptedStorage: EncryptedStorageType & LogOutHandler
     private let localStorage: LocalStorageType & LogOutHandler
-    private let dataService: DataServiceType
 
     private let imap: Imap
     private let googleService: GoogleUserService
+    private let dataService: DataService
```
Comment (tomholub) [inline FlowCrypt/Functionality/DataManager/SessionService.swift:26]: Should be `DataServiceType`, I suppose

File: FlowCrypt/Functionality/DataManager/SessionService.swift [hunk 16]
```diff
@@ -94,21 +93,16 @@ extension UserAccountService: UserAccountServiceType {
         return switchActiveSession(for: currentUser)
     }
 
+    // todo - rename to "logOutUsersThatDontHaveAnyKeysSetUp"
     func cleanupSessions() {
```
Comment (tomholub) [inline FlowCrypt/Functionality/DataManager/SessionService.swift:97]: Let's do this rename

File: FlowCrypt/Functionality/Mail Provider/Gmail/GmailService.swift [hunk 17]
```diff
@@ -35,7 +36,8 @@ class GmailService: MailServiceProvider {
     var progressHandler: ((Float) -> Void)?
 
     init(
-        userService: GoogleUserServiceType = GoogleUserService(),
+        currentUserEmail: String,
+        userService: GoogleUserServiceType,
```
Comment (tomholub) [inline FlowCrypt/Functionality/Mail Provider/Gmail/GmailService.swift:40]: Let's call this property `googleUserService`

File: FlowCrypt/Functionality/Services/Client Configuration Service/ClientConfigurationService.swift [hunk 18]
```diff
@@ -10,51 +10,48 @@ import FlowCryptCommon
...
 extension ClientConfigurationService: ClientConfigurationServiceType {
 
-    func fetchForCurrentUser() async throws -> ClientConfiguration {
-        guard let currentUserEmail = getCurrentUserEmail() else {
-            throw AppErr.noCurrentUser
-        }
+    func fetch(for user: User) async throws -> ClientConfiguration {
+//        guard let user = user else {
+//            throw AppErr.noCurrentUser
+//        }
```
Comment (tomholub) [inline FlowCrypt/Functionality/Services/Client Configuration Service/ClientConfigurationService.swift:37]: These can be removed

File: FlowCrypt/Functionality/Services/Folders Services/TrashFolderProvider.swift [hunk 19]
```diff
@@ -9,13 +9,17 @@
 struct TrashFolderProvider {
     private let localStorage: LocalStorageType
     private let folderProvider: FoldersServiceType
+    private let user: User
 
     init(
-        folderProvider: FoldersServiceType = FoldersService(),
+        // todo - rename argument to folderService:
+        user: User,
+        folderProvider: FoldersServiceType,
```
Comment (tomholub) [inline FlowCrypt/Functionality/Services/Folders Services/TrashFolderProvider.swift:17]: rename `folderProvider` to `folderService` here

File: FlowCrypt/Functionality/Services/GlobalRouter.swift [hunk 20]
```diff
@@ -114,45 +121,40 @@ extension GlobalRouter: GlobalRouterType {
...
+        keyWindow.rootViewController = UINavigationController(rootViewController: controller)
+        keyWindow.makeKeyAndVisible()
     }
 
     @MainActor
-    private func proceed(with session: SessionType?) {
-        logger.logInfo("proceed for session \(session.debugDescription)")
-        AppStartup().initializeApp(window: keyWindow, session: session)
+    private func proceed(with appContext: AppContext) {
+        logger.logInfo("proceed for session: \(appContext.session?.description ?? "nil")")
```
Comment (tomholub) [inline FlowCrypt/Functionality/Services/GlobalRouter.swift:147]: originally it used `debugDescription`, not sure of the difference

File: FlowCrypt/Functionality/Services/Local Private Key Services/KeyDataStorage.swift [hunk 21]
```diff
@@ -9,13 +9,12 @@
 import FlowCryptCommon
 import Foundation
 
+// todo - what is this and what is it used for?
 final class KeyDataStorage {
```
Comment (tomholub) [inline FlowCrypt/Functionality/Services/Local Private Key Services/KeyDataStorage.swift:13]: Please remove this class, it just wraps functionality with no value add

File: FlowCrypt/Functionality/Services/Local Private Key Services/KeyService.swift [hunk 22]
```diff
@@ -28,9 +28,9 @@ final class KeyService: KeyServiceType {
     let logger: Logger
 
     init(
-        storage: KeyStorageType = KeyDataStorage(),
-        passPhraseService: PassPhraseServiceType = PassPhraseService(),
-        currentUserEmail: @autoclosure @escaping () -> (String?) = DataService.shared.email
+        storage: KeyStorageType,
```
Comment (tomholub) [inline FlowCrypt/Functionality/Services/Local Private Key Services/KeyService.swift:31]: please rename `storage` to `keyStorage` here

File: FlowCrypt/Functionality/Services/Local Private Key Services/PassPhraseService.swift [hunk 23]
```diff
@@ -72,18 +72,15 @@ protocol PassPhraseServiceType {
...
     let encryptedStorage: PassPhraseStorageType
     let inMemoryStorage: PassPhraseStorageType
 
     init(
-        encryptedStorage: PassPhraseStorageType = EncryptedStorage(),
-        localStorage: PassPhraseStorageType = InMemoryPassPhraseStorage(),
-        emailProvider: EmailProviderType = DataService.shared
+        encryptedStorage: PassPhraseStorageType,
+        localStorage: PassPhraseStorageType = InMemoryPassPhraseStorage()
     ) {
```
Comment (tomholub) [inline FlowCrypt/Functionality/Services/Local Private Key Services/PassPhraseService.swift:81]: This naming is extremely confusing an we'll need to file a separate issue just for this

**Review by tomholub — COMMENTED** [review]
Below are smaller cleanup/refactors that I wanted to do in this PR
